hedera-node/src/main/java/com/hedera/services/store/tokens/unique/UniqueTokenStore.java

```diff
@@ -0,0 +1,196 @@
+			final var suppliedTokens = uniqueTokensSupply.get();
+			final var eId = EntityId.fromGrpcTokenId(tId);
+			final var owner = get(tId).treasury(); // get next available serial num here as well
+			final int serialNum = 0;
+			// TODO serialNum
+			// When a merkleUniqueToken is created, the next present serial number must be available as a method
+			final var nftId = new MerkleUniqueTokenId(eId, serialNum);
+			final var nft = new MerkleUniqueToken(owner, memo, creationTime);
+			final var putResult = suppliedTokens.putIfAbsent(nftId, nft);
+			return (putResult == null && mintResult == ResponseCodeEnum.OK) ? ResponseCodeEnum.OK : ResponseCodeEnum.INVALID_TOKEN_ID;
+		}));
+
+	}
+
+	@Override
+	public MerkleUniqueToken getUnique(final EntityId eId, final int serialNum) {
+		return uniqueTokensSupply.get().get(new MerkleUniqueTokenId(eId, serialNum));
+	}
+
+	@Override
+	public Iterator<MerkleUniqueTokenId> getByToken(final MerkleUniqueToken token) {
+		return uniqueTokensSupply.get().inverseGet(token);
+	}
+
+	@Override
+	public Iterator<MerkleUniqueTokenId> getByTokenFromIdx(final MerkleUniqueToken token, final int start) {
+		if(start > uniqueTokensSupply.get().size()){
+			throw new IllegalArgumentException("Start index " + start + " is larger than size.");
+		}
+		if (start < 0){
+			throw new IllegalArgumentException("Start index cannot be negative");
+		}
+
+		return uniqueTokensSupply.get().inverseGet(token, start);
+	}
+
+	@Override
+	public Iterator<MerkleUniqueTokenId> getByTokenFromIdxToIdx(final MerkleUniqueToken token, final int start, final int end) {
+		if(start > uniqueTokensSupply.get().size())
+			throw new IllegalArgumentException("Start index " + start + " is larger than size.");
+
+		if (start < 0)
+			throw new IllegalArgumentException("Start index cannot be negative.");
+
+		if(end > uniqueTokensSupply.get().size())
+			throw new IllegalArgumentException("End index "+ end + " is larger than size.");
+
+		if (end < 0)
+			throw new IllegalArgumentException("End index cannot be negative.");
+
+		if (end < start)
+			throw new IllegalArgumentException("Start index cannot be bigger than end. "+ start + " > "+ end);
+
+		return uniqueTokensSupply.get().inverseGet(token, start, end);
+	}
+
+	// currently not taking care of start/end
+	@Override
+	public Iterator<MerkleUniqueTokenId> getByAccountFromIdxToIdx(final AccountID aId, final int start, final int end) {
+
+		if(end < 0 || start < 0)
+			throw new IllegalArgumentException("Start or End index is negative. s- " + start + " e - "+ end);
+
+		if(end < start)
+			throw new IllegalArgumentException("Start index cannot be bigger than end. "+ start + " > "+ end);
+
+		if (end > uniqueTokensSupply.get().size())
+			throw new IllegalArgumentException("End index "+ end + " is larger than size.");
+
+		final var eId = EntityId.fromGrpcAccountId(aId);
+		var tokenSupply = uniqueTokensSupply.get();
+		final var ownerIdentifier = new OwnerIdentifier(eId);
+		return tokenSupply
```

Comment:
`inverseGet` does the same logic more effectively. We should be using that functionality.